File: sdk/kotlin/src/main/kotlin/com/telecom/AnalyticsAPI.kt
package com.telecom

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonObject

/**
 * Analytics API for churn prediction, market analysis, and pricing optimization
 */
class AnalyticsAPI(private val httpClient: HTTPClient) {
    
    /**
     * Predict churn risk for a profile
     */
    suspend fun predictChurn(profileId: String): ChurnPrediction {
        return httpClient.post("/api/v1/analytics/churn/predict", 
            mapOf("profile_id" to profileId))
    }
    
    /**
     * Get churn metrics
     */
    suspend fun getChurnMetrics(period: String = "monthly"): ChurnMetrics {
        return httpClient.get("/api/v1/analytics/churn/metrics", 
            mapOf("period" to period))
    }
    
    /**
     * Get at-risk customers
     */
    suspend fun getAtRiskCustomers(
        riskLevel: ChurnRiskLevel, 
        limit: Int = 100
    ): List<ChurnPrediction> {
        return httpClient.post("/api/v1/analytics/churn/at-risk",
            mapOf(
                "risk_level" to riskLevel.value,
                "limit" to limit
            ))
    }
    
    /**
     * Get market metrics
     */
    suspend fun getMarketMetrics(period: String = "monthly"): MarketMetrics {
        return httpClient.get("/api/v1/analytics/market/metrics",
            mapOf("period" to period))
    }
    
    /**
     * Get competitor analysis
     */
    suspend fun getCompetitors(): JsonObject {
        return httpClient.get("/api/v1/analytics/market/competitors")
    }
    
    /**
     * Get market opportunities
     */
    suspend fun getMarketOpportunities(): JsonObject {
        return httpClient.get("/api/v1/analytics/market/opportunities")
    }
    
    /**
     * Get maintenance metrics
     */
    suspend fun getMaintenanceMetrics(period: String = "monthly"): MaintenanceMetrics {
        return httpClient.get("/api/v1/analytics/maintenance/metrics",
            mapOf("period" to period))
    }
    
    /**
     * Get assets health
     */
    suspend fun getAssetsHealth(): JsonObject {
        return httpClient.get("/api/v1/analytics/maintenance/assets")
    }
    
    /**
     * Get maintenance alerts
     */
    suspend fun getMaintenanceAlerts(): JsonObject {
        return httpClient.get("/api/v1/analytics/maintenance/alerts")
    }
    
    /**
     * Predict failure for an asset
     */
    suspend fun predictFailure(assetId: String): JsonObject {
        return httpClient.post("/api/v1/analytics/maintenance/predict/$assetId", emptyMap())
    }
    
    /**
     * Get pricing metrics
     */
    suspend fun getPricingMetrics(period: String = "monthly"): PricingMetrics {
        return httpClient.get("/api/v1/analytics/pricing/metrics",
            mapOf("period" to period))
    }
    
    /**
     * Optimize pricing for rate plans
     */
    suspend fun optimizePricing(
        ratePlanIds: List<String>,
        strategy: String = "revenue_maximization"
    ): List<PricingOptimizationResult> {
        return httpClient.post("/api/v1/analytics/pricing/optimize",
            mapOf(
                "rate_plan_ids" to ratePlanIds,
                "strategy" to strategy
            ))
    }
    
    /**
     * Get price elasticity data
     */
    suspend fun getPriceElasticity(): JsonObject {
        return httpClient.get("/api/v1/analytics/pricing/elasticity")
    }
}

@Serializable
data class ChurnPrediction(
    val profileId: String,
    val riskLevel: String,
    val riskScore: Double,
    val predictedChurnDate: String? = null,
    val reasons: List<String>,
    val recommendations: List<String>,
    val lastUpdated: String
)

@Serializable
data class ChurnMetrics(
    val period: String,
    val totalSubscribers: Long,
    val churnedSubscribers: Long,
    val churnRate: Double,
    val monthlyChurnRate: Double,
    val annualChurnRate: Double,
    val averageTenureDays: Double,
    val riskDistribution: Map<String, Long>,
    val generatedAt: String
)

@Serializable
data class MarketMetrics(
    val period: String,
    val totalMarketSize: Long,
    val ourSubscribers: Long,
    val marketShare: Double,
    val growthRate: Double,
    val byCountry: Map<String, JsonObject>,
    val generatedAt: String
)

@Serializable
data class MaintenanceMetrics(
    val period: String,
    val totalAssets: Long,
    val healthyAssets: Long,
    val assetsNeedingAttention: Long,
    val uptime: Double,
    val meanTimeToFailure: Double,
    val meanTimeToRepair: Double,
    val generatedAt: String
)

@Serializable
data class PricingMetrics(
    val period: String,
    val totalRevenue: Double,
    val arpu: Double,
    val priceElasticity: Double,
    val competitiveIndex: Double,
    val optimizationRoi: Double,
    val generatedAt: String
)

@Serializable
data class PricingOptimizationResult(
    val ratePlanId: String,
    val strategy: String,
    val currentPrice: Double,
    val optimalPrice: Double,
    val priceChangePct: Double,
    val expectedRevenue: Double,
    val expectedDemand: Double,
    val confidence: Double,
    val reasoning: List<String>,
    val risks: List<String>,
    val recommendations: List<String>
)

@Serializable
enum class ChurnRiskLevel(val value: String) {
    LOW("low"),
    MEDIUM("medium"),
    HIGH("high"),
    CRITICAL("critical")
}

File: sdk/swift/Sources/TelecomSDK/AnalyticsAPI.swift
import Foundation

/// Analytics API for churn prediction, market analysis, and pricing optimization
public class AnalyticsAPI {
    private let client: HTTPClient
    
    public init(client: HTTPClient) {
        self.client = client
    }
    
    // MARK: - Churn Analysis
    
    /// Predict churn risk for a profile
    public func predictChurn(profileId: String) async throws -> ChurnPrediction {
        return try await client.post(
            "/api/v1/analytics/churn/predict",
            body: ["profile_id": profileId],
            responseType: ChurnPrediction.self
        )
    }
    
    /// Get churn metrics
    public func getChurnMetrics(period: String = "monthly") async throws -> ChurnMetrics {
        return try await client.get(
            "/api/v1/analytics/churn/metrics",
            parameters: ["period": period],
            responseType: ChurnMetrics.self
        )
    }
    
    /// Get at-risk customers
    public func getAtRiskCustomers(
        riskLevel: ChurnRiskLevel,
        limit: Int = 100
    ) async throws -> [ChurnPrediction] {
        return try await client.post(
            "/api/v1/analytics/churn/at-risk",
            body: [
                "risk_level": riskLevel.rawValue,
                "limit": limit
            ],
            responseType: [ChurnPrediction].self
        )
    }
    
    // MARK: - Market Analytics
    
    /// Get market metrics
    public func getMarketMetrics(period: String = "monthly") async throws -> MarketMetrics {
        return try await client.get(
            "/api/v1/analytics/market/metrics",
            parameters: ["period": period],
            responseType: MarketMetrics.self
        )
    }
    
    /// Get competitor analysis
    public func getCompetitors() async throws -> [String: Any] {
        return try await client.get("/api/v1/analytics/market/competitors")
    }
    
    /// Get market opportunities
    public func getMarketOpportunities() async throws -> [String: Any] {
        return try await client.get("/api/v1/analytics/market/opportunities")
    }
    
    // MARK: - Predictive Maintenance
    
    /// Get maintenance metrics
    public func getMaintenanceMetrics(period: String = "monthly") async throws -> MaintenanceMetrics {
        return try await client.get(
            "/api/v1/analytics/maintenance/metrics",
            parameters: ["period": period],
            responseType: MaintenanceMetrics.self
        )
    }
    
    /// Get assets health
    public func getAssetsHealth() async throws -> [String: Any] {
        return try await client.get("/api/v1/analytics/maintenance/assets")
    }
    
    /// Get maintenance alerts
    public func getMaintenanceAlerts() async throws -> [String: Any] {
        return try await client.get("/api/v1/analytics/maintenance/alerts")
    }
    
    /// Predict failure for an asset
    public func predictFailure(assetId: String) async throws -> [String: Any] {
        return try await client.post("/api/v1/analytics/maintenance/predict/\(assetId)", body: [:])
    }
    
    // MARK: - Pricing Optimization
    
    /// Get pricing metrics
    public func getPricingMetrics(period: String = "monthly") async throws -> PricingMetrics {
        return try await client.get(
            "/api/v1/analytics/pricing/metrics",
            parameters: ["period": period],
            responseType: PricingMetrics.self
        )
    }
    
    /// Optimize pricing for rate plans
    public func optimizePricing(
        ratePlanIds: [String],
        strategy: String = "revenue_maximization"
    ) async throws -> [PricingOptimizationResult] {
        return try await client.post(
            "/api/v1/analytics/pricing/optimize",
            body: [
                "rate_plan_ids": ratePlanIds,
                "strategy": strategy
            ],
            responseType: [PricingOptimizationResult].self
        )
    }
    
    /// Get price elasticity data
    public func getPriceElasticity() async throws -> [String: Any] {
        return try await client.get("/api/v1/analytics/pricing/elasticity")
    }
}

// MARK: - Analytics Types

public struct ChurnPrediction: Codable {
    public let profileId: String
    public let riskLevel: String
    public let riskScore: Double
    public let predictedChurnDate: String?
    public let reasons: [String]
    public let recommendations: [String]
    public let lastUpdated: String
}

public struct ChurnMetrics: Codable {
    public let period: String
    public let totalSubscribers: Int64
    public let churnedSubscribers: Int64
    public let churnRate: Double
    public let monthlyChurnRate: Double
    public let annualChurnRate: Double
    public let averageTenureDays: Double
    public let riskDistribution: [String: Int64]
    public let generatedAt: String
}

public struct MarketMetrics: Codable {
    public let period: String
    public let totalMarketSize: Int64
    public let ourSubscribers: Int64
    public let marketShare: Double
    public let growthRate: Double
    public let byCountry: [String: [String: Any]]
    public let generatedAt: String
}

public struct MaintenanceMetrics: Codable {
    public let period: String
    public let totalAssets: Int64
    public let healthyAssets: Int64
    public let assetsNeedingAttention: Int64
    public let uptime: Double
    public let meanTimeToFailure: Double
    public let meanTimeToRepair: Double
    public let generatedAt: String
}

public struct PricingMetrics: Codable {
    public let period: String
    public let totalRevenue: Double
    public let arpu: Double
    public let priceElasticity: Double
    public let competitiveIndex: Double
    public let optimizationRoi: Double
    public let generatedAt: String
}

public struct PricingOptimizationResult: Codable {
    public let ratePlanId: String
    public let strategy: String
    public let currentPrice: Double
    public let optimalPrice: Double
    public let priceChangePct: Double
    public let expectedRevenue: Double
    public let expectedDemand: Double
    public let confidence: Double
    public let reasoning: [String]
    public let risks: [String]
    public let recommendations: [String]
}

public enum ChurnRiskLevel: String, CaseIterable {
    case low = "low"
    case medium = "medium"
    case high = "high"
    case critical = "critical"
}
